trie/triedb/pathdb/history.go

```diff
@@ -482,13 +494,93 @@ func (h *history) decode(accountData, storageData, accountIndexes, storageIndexe
 			storages[accIndex.address] = slotData
 		}
 	}
+
+	if err := rlp.DecodeBytes(trieNodes, &encodedTrieNodes); err != nil {
+		log.Error("Failed to decode state history trie nodes", "error", err)
+		return err
+	}
+
 	h.accounts = accounts
 	h.accountList = accountList
 	h.storages = storages
 	h.storageList = storageList
+	h.nodes = encodedTrieNodes
 	return nil
 }
 
+func (h *history) Size() int {
+	size := 0
+
+	// calculate size of meta
+	if h.meta != nil {
+		metaSize := 1
+		metaSize += len(h.meta.parent) + len(h.meta.root)
+		metaSize += 8
+		metaSize += len(h.meta.incomplete) * (common.AddressLength)
+
+		size += metaSize
+	}
+	// calculate size of accounts
+	for address, data := range h.accounts {
+		size += len(address)
+		size += len(data)
+	}
+	// calculate size of accountList
+	size += len(h.accountList) * common.AddressLength
+	// calculate size of storages
+	for _, storage := range h.storages {
+		for slotHash, data := range storage {
+			size += len(slotHash)
+			size += len(data)
+		}
+	}
+	// calculate size of storageList
+	for _, slots := range h.storageList {
+		size += len(slots) * common.HashLength
+	}
+	// calculate size of journalNodes
+	for _, jn := range h.nodes {
+		size += common.AddressLength
+		for _, n := range jn.Nodes {
+			size += len(n.Path)
+			size += len(n.Blob)
+		}
+	}
+
+	return size
+}
+
+// readBlockNumber reads and decodes the state history meta and returns block number.
+func readBlockNumber(freezer *rawdb.ResettableFreezer, stateID uint64) (uint64, error) {
+	blob := rawdb.ReadStateHistoryMeta(freezer, stateID)
+	if len(blob) == 0 {
+		return 0, fmt.Errorf("state history not found %d", stateID)
+	}
+	var m meta
+	if err := m.decode(blob); err != nil {
+		return 0, err
+	}
+	return m.block, nil
+}
+
+// readAllBlockNumbers returns all block number to stateID map.
+func readAllBlockNumbers(freezer *rawdb.ResettableFreezer, startStateID, endStateID uint64) (map[uint64]uint64, error) {
+	blockMap := make(map[uint64]uint64)
+	for i := startStateID; i <= endStateID; i++ {
+		blob := rawdb.ReadStateHistoryMeta(freezer, i)
+		if len(blob) == 0 {
+			return nil, fmt.Errorf("state history not found %d", i)
+		}
+		var m meta
+		if err := m.decode(blob); err != nil {
+			return nil, err
+		}
+		blockMap[m.block] = i
+	}
+	log.Info("state id", "startStateID", startStateID, "endStateID", endStateID, "length", len(blockMap))
```

Comment:
Refine  the`state id` msg to keep meanful.

Reply:
Deleted